docs/autodiff.rst

```diff
@@ -0,0 +1,239 @@
+.. _differentiable:
+
+Differentiable programming
+==========================
+
+We suggest starting with the ``ti.Tape()``, and then migrate to more advanced differentiable programming using the ``kernel.grad()`` syntax if necessary.
+
+
+Introduction
+------------
+
+For example, you have the following kernel:
+
+.. code-block:: python
+
+    x = ti.var(ti.f32, ())
+    y = ti.var(ti.f32, ())
+
+    @ti.kernel
+    def compute_y():
+        y[None] = ti.sin(x[None])
+
+
+Now you want to get the derivative of y corresponding to x, i.e. dy/dx.
+You may want to write the derivatived kernel by hand:
+
+.. code-block:: python
+
+    x = ti.var(ti.f32, ())
+    y = ti.var(ti.f32, ())
+    dy_dx = ti.var(ti.f32, ())
+
+    @ti.kernel
+    def compute_dy_dx():
+        dy_dx[None] = ti.cos(x[None])
+
+
+But wait, what if I changed the original ``compute_y``? We will have to
+recalculate the derivative by hand and rewrite ``compute_dy_dx`` again, which
+is very error-prone and not convienent at all.
+
+If you meet this situation, don't worry! Taichi provides a handy autodiff
+system that can help you obtain the derivative of a kernel without any pain!
+
+
+Using ``ti.Tape()``
+-------------------
+
+Let's still take the ``compute_y`` in above example for explaination.
+What's the most convienent way to obtain a kernel that computes x to dy/dx?
+
+1. Use the ``needs_grad=True`` option when declaring tensors involved in the
+   derivative chain.
+2. Use ``with ti.Tape(y):`` to enbrace the invocation into kernel(s) you want
+   to compute derivative.
```

Comment:
IIUIC this means we use `with.ti.Tape(y)` to allow the invocation into kernels? 🤔 

```suggestion
2. Use ``with ti.Tape(y):`` to embrace the invocation into kernel(s) you want
   to compute derivative.
```